### ISSUE TITLE
Specify an FPS limit for interpolation statements in ATL

### ISSUE THREAD

**Issue by colinharvie** [renpy/renpy#5463]
For stylistic reasons, I want to be able to restrict interpolations to specific frame rates (globally, as well as per interpolation). I tend to use a mix of 8, 12, and 24 for various different things (and I can see use cases for other frame rates).

The current way I do this is with a curried function:

```
def _fpsWarp(fps=24,dur=0,warper="linear",t=0):
    if isinstance(warper,str) and hasattr(_warper,warper):
        warper = getattr(_warper,warper)
    elif not hasattr(warper,'__call__'):
        warper = _warper.linear
    t = warper(t)
    return floor(t*dur*fps)/fps/dur

fpsWarp = renpy.curry(_fpsWarp)
```

Which I call like so:

```
scene bg side_hallway:
    xalign 0.5 yalign 0.5
    block:
        xpan 0
        warp fpsWarp(24,12.5,"linear") 12.5 xpan 360
        repeat
```
This lets me limit the fps of the interpolation xpan to 24fps (or whatever I want), instead of 30/60 whatever my screen is doing.
I want this for stylistic reasons; my whole game is designed to look like an animated cartoon, and having some interpolations move at 30/60fps can really mess with the eyes. 

Now, this works fine for me, BUT there are a few things that make it a bit cumbersome:
1 - It's a lot more text every time I want to set the frame rate on an interpolation.
2 - I have to specify the the duration twice, since warpers only get (t) as a percentage of completion.

It would be very useful to have the following:

- way to set a global framerate for interpolations in config (None or False could be unlimited/as current)
- ATL statement to set the framerate for the current Transform (or current block?)
- Interpolation statements can take an additional integer (or variable), that specifies the frame rate to limit that specific statement to.

Something like this:

```
#sets the global interpolation framerate to 24fps
$ config.interp_framerate = 24 

transform test_atl():
    #set base framerate for this whole transform, otherwise transform uses config.interp_framerate
    framerate 12 

    xalign 0.0

    #no specified framrate, will use the one specified above (12)
    ease 1.0 xalign 1.0 

    #second variable after duration, sets the framerate for this interpolation 
    easein 1.0 24 xalign 0.0 
    
    #second variable is None, no framerate limit
    linear 1.0 None xalign 0.0 
```

I imagine this won't be something EVERYONE needs, but for those of us that want that little stylistic kick, it'd be really nice and helpful (and make my ATL code look a lot cleaner XD)

**Comment by Gouvernathor** [comment]
I don't see this happening for solely stylistic reasons, as it would complexify interpolation code a lot (when it's already quite intricate and pending [an update](https://github.com/renpy/renpy/tree/gouv-display-type-improvements)) for little use and while you already have a way around it.
Interesting solution though.

**Comment by renpytom** [comment]
I'm not sure a per-interpolation or per-ATL statement is worth having, but you can set preferences.gl_framerate to globally adjust things.

**Comment by colinharvie** [comment]
I did initially try using `preferences.gl_framerate` for an overall limitation, but the problem with that was that it made certain things feel real sluggish (ui interaction, custom mouse cursor, things of that nature).

I did manage to make a bit of a hack that gives me way more control over things, though it's definitely modifying bits of the code that make me nervous and I'll have to keep an eye on any engine updates that touch this area, but it works quite well! 

First, I fixed up the code from above first, adding in conditions to handle if fps is None or if duration was 0

```
def _fpsWarp(fps=24,dur=0,warper="linear",t=0):
    warper = getattr(_warper,warper,warper)
    if not hasattr(warper,'__call__'):
        warper = _warper.linear
    t=warper(t)

    if fps is None or dur == 0:
        return t
    
    return floor(t*dur*fps)/(fps*dur)

fpsWarp = renpy.curry(_fpsWarp)
```
Then, I add tell ATL that "fps" is a valid property >_>
`renpy.display.transform.add_property("fps")`

And, uh, here's where it gets real sketchy, I replace the `__init__` method of renpy.atl.Interpolation

```
def _Interpolation__init__(self, loc, warper, duration, properties, revolution, circles, splines):
    super(renpy.atl.Interpolation, self).__init__(loc)

    self.warper = warper
    self.duration = duration
    self.properties = properties
    self.splines = splines

    # The direction we revolve in: cw, ccw, or None.
    self.revolution = revolution

    # The number of complete circles we make.
    self.circles = circles

    #check for FPS property, add fps limit warper if exists
    fps = [x[1] for x in properties if x[0] == "fps"]
    if duration and len(fps):
        self.warper = fpsWarp(fps[0],duration,warper)

renpy.atl.Interpolation.__init__ = _Interpolation__init__
```
I basically modify the warper in Interpolation creation, telling it to use one of my fpsWarps, IF it finds an FPS property on the
So now, if I want an interpolation with a specific FPS, all I gotta do is:

```
show test_img:
    xpos 0
    ease 3.0 fps 8 xpos 1.0
```    

And now it runs at 8fps, (or whatever set the fps property to) with whatever warper I give it!
It hasn't crashed on me yet!

**Comment by renpytom** [comment]
I've added an fps property to Transform that does something very similar at a different level, so hopefully you'll get what you want. I'll be in the April 13 nightly build.

**Comment by colinharvie** [comment]
I checked it out and did a few tests! For the most part it seems to work, but I've also immediately broken it. I don't know WHY this breaks it, but I ahve the following setup:

A "wall" background, which has a repeating xpan at 24 fps. Characters animated on top of that background (with a sequence of pauses and properties) , and finally a full black solid on a custom "cvr" layer, which slowly fades out, revealing them. During that fade transition, the FPS is very strange,

Here are two video files showing the problem. One is at 24fps (my intended use), and you can see the strange stuttering / fps ramping as everything fades in. The second is at 4fps, which just lets you see the stuttering more clearly. (It actually looks much more pronounced in game, with flickering and "ghosting", my video recorder just maxes out at 30fps.)

24fps, 8,3 24041301+nightly
https://github.com/renpy/renpy/assets/4917213/45aeb482-b146-4a4c-9b8c-a2dad6d31ca3

4fps, 8,3 24041301+nightly
https://github.com/renpy/renpy/assets/4917213/e5a2b8e6-0934-44aa-b068-816263524c27

For comparison, here are recordings on the same from 8.2.1, using the hack I mentioned above:

24fps, 8,2.1.24030407
https://github.com/renpy/renpy/assets/4917213/e191c9b6-76c6-48b5-9dff-43b86df86788

4fps, 8,2.1.24030407
https://github.com/renpy/renpy/assets/4917213/ab2b14c1-34bd-4ba9-98ef-0fbcf0c8bdce



I also looked at the code, and to be honest, what's going on there is a bit beyond my comprehension, but wouldn't calculating the frame limit BEFORE applying the warper cause some issues? (It might not, I haven't done any testing yet, and as I said the way it's being calculated here is beyond me).

**Comment by colinharvie** [comment]
I'm a dummy, I figured out what I've done wrong.

I still had the `fps` property in the interpolation statement, and I've just realized what I was seeing was watching it interpolate from 0 fps to 24 fps. 

Moved the `fps` property out of the interpolation statement, and it works like a treat! 

The only I'd add here is that this method doesn't let you mix FPSs if you have multiple interpolations going with some parallel blocks. If it could work per statement block (not just per transform) that'd be great, BUT honestly that's a much more specific and infrequent need, and I could use a `function` atl statement in those rare cases.  

So works great! Once I figured out what I'd done wrong.
